{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from nanopore_tcr_consensus.analysis import (\n",
    "    set_mpl_params,\n",
    "    split_ref_names_set_into_subsets,\n",
    "    parse_raw_nanopore_qual_from_fastq_stats,\n",
    "    parse_quantile_95_blast_id_from_self_homology_log,\n",
    "    parse_merged_consensus_bam_filter_log,\n",
    "    plot_blast_id_with_most_similar_tcr,\n",
    "    plot_hist_subreads_per_umi_cluster,\n",
    "    plot_box_blast_id_vs_subreads_umi_cluster,\n",
    "    plot_hist_nt_too_short,\n",
    "    plot_hist_nt_too_long,\n",
    "    plot_hist_percent_alignments_above_blast_id,\n",
    "    log_normalize_umi_counts,\n",
    "    plot_umi_counts_hist,\n",
    "    plot_log_transformed_umi_counts_hist,\n",
    "    plot_v_gene_fold_change_over_input_barplot,\n",
    "    plot_v_gene_fraction_missing_tcrs,\n",
    "    filter_counts_on_log_umi_count_threshold,\n",
    "    write_results_summary_txt,\n",
    "    plot_plate_umi_counts,\n",
    ")\n",
    "from nanopore_tcr_consensus.region_split import generate_regions_set_from_ref_fa\n",
    "\n",
    "import os\n",
    "import pandas as pd\n",
    "import glob\n",
    "import re\n",
    "\n",
    "import matplotlib as mpl\n",
    "from matplotlib.cm import viridis\n",
    "\n",
    "viridis.set_bad(\"lightgrey\")\n",
    "mpl, mylines = set_mpl_params(mpl=mpl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nanopore_project_dir = \"\"\n",
    "nano_tcr_name = \"nano_tcr\"\n",
    "internal_basecalling = True\n",
    "plot_plates = False\n",
    "\n",
    "blast_id_ylim_max = 350\n",
    "umi_count_xlim_max = 900\n",
    "umi_count_bin_size = 10\n",
    "umi_count_ylim_max = 0.1\n",
    "log_umi_count_xlim_max = 8\n",
    "log_umi_count_bin_size = 0.25\n",
    "log_umi_count_ylim_max = 1.0\n",
    "most_similar_blast_id_threshold = 0.99925\n",
    "\n",
    "\n",
    "libraries_csv = os.path.join(nanopore_project_dir, \"libraries.csv\")\n",
    "nanopore_project_outs_dir = os.path.join(nanopore_project_dir, \"outs\")\n",
    "os.mkdir(nanopore_project_outs_dir)\n",
    "libraries_df = pd.read_csv(libraries_csv)\n",
    "libraries_df.to_csv(os.path.join(nanopore_project_outs_dir, \"libraries.csv\"))\n",
    "library_list = libraries_df.loc[:, \"barcode\"].tolist()\n",
    "if internal_basecalling:\n",
    "    library_list = [\"concatenated_\" + library for library in library_list]\n",
    "library_name_list = libraries_df.loc[:, \"library_name\"].tolist()\n",
    "if any([\" \" in library_name for library_name in library_name_list]):\n",
    "    raise Exception(\"white space is not allowed in library names!\")\n",
    "library_ref_id_list = libraries_df.loc[:, \"ref_library_name\"].tolist()\n",
    "library_ref_id_list = [\n",
    "    False if pd.isna(library_ref_id) else library_ref_id\n",
    "    for library_ref_id in library_ref_id_list\n",
    "]\n",
    "log_umi_counts_filter_threshold_list = libraries_df.loc[\n",
    "    :, \"log_umi_counts_filter_threshold\"\n",
    "].tolist()\n",
    "if len(glob.glob(os.path.join(nanopore_project_dir, \"*.fa\"))) > 1:\n",
    "    raise Exception(\n",
    "        \"There is more than reference .fa file in the nanopore_project_dir:\",\n",
    "        nanopore_project_dir,\n",
    "    )\n",
    "reference_name = glob.glob(os.path.join(nanopore_project_dir, \"*.fa\"))[0].split(\".fa\")[\n",
    "    0\n",
    "]\n",
    "reference = os.path.join(nanopore_project_dir, reference_name + \".fa\")\n",
    "ref_names_set = generate_regions_set_from_ref_fa(reference=reference)\n",
    "non_n_ref_names_set, full_n_ref_names_set, cdr3j_n_ref_names_set, v_n_ref_names_set = (\n",
    "    split_ref_names_set_into_subsets(ref_names_set=ref_names_set)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# custom_tcr_set_dict = None\n",
    "gil_plate_1_duplicate_1_ref_set = set()\n",
    "pattern = r\"^[A-H](?:[1-9]|1[0-9]|2[0-4])$\"\n",
    "for ref in non_n_ref_names_set:\n",
    "    if (\n",
    "        ref.endswith(\"GILGFVFTL\")\n",
    "        and ref.startswith(\"1_1\")\n",
    "        and re.match(pattern, ref.split(\"_\")[2])\n",
    "    ):\n",
    "        gil_plate_1_duplicate_1_ref_set.add(ref)\n",
    "\n",
    "custom_tcr_set_dict = {\"gil_plate_1_tcrs\": gil_plate_1_duplicate_1_ref_set}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing: vdjdb_technical_duplicate_1\n",
      "Plotting blast_id_with_most_similar_tcr\n",
      "Plotting hist_subreads_per_umi_cluster\n",
      "Plotting box_blast_id_vs_subreads_umi_cluster\n",
      "Plotting hist_nt_too_short\n",
      "Plotting hist_nt_too_long\n",
      "Plotting hist_percent_alignments_above_blast_id\n",
      "Plotting umi_counts_hists\n",
      "Warning: xlim_max is smaller than the 99th percentile UMI count: 1121.1499999999992\n",
      "KS Statistic: 0.09697589911308102, p-value: 2.1457867866486665e-30\n",
      "Warning: xlim_max is smaller than the 99th percentile UMI count: 1121.1499999999992\n",
      "Plotting log_transformed_umi_counts_hists\n",
      "KS Statistic: 0.20663946070371503, p-value: 1.8240884995187867e-137\n",
      "Plotting v_gene_fold_change_over_input_barplot\n",
      "Plotting v_gene_fraction_missing_tcrs\n",
      "Writing results_summary_txt\n",
      "Writing filtered_umi_count csv\n",
      "\n",
      "\n",
      "\n",
      "Processing: vdjdb_technical_duplicate_2\n",
      "Plotting blast_id_with_most_similar_tcr\n",
      "Plotting hist_subreads_per_umi_cluster\n",
      "Plotting box_blast_id_vs_subreads_umi_cluster\n",
      "Plotting hist_nt_too_short\n",
      "Plotting hist_nt_too_long\n",
      "Plotting hist_percent_alignments_above_blast_id\n",
      "Plotting umi_counts_hists\n",
      "Warning: xlim_max is smaller than the 99th percentile UMI count: 1245.9799999999996\n",
      "KS Statistic: 0.10598730428168571, p-value: 7.629108194013631e-36\n",
      "Warning: xlim_max is smaller than the 99th percentile UMI count: 1245.9799999999996\n",
      "Plotting log_transformed_umi_counts_hists\n",
      "KS Statistic: 0.12453133792794241, p-value: 2.309675264616618e-49\n",
      "Plotting v_gene_fold_change_over_input_barplot\n",
      "Plotting v_gene_fraction_missing_tcrs\n",
      "Writing results_summary_txt\n",
      "Writing filtered_umi_count csv\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for (\n",
    "    library,\n",
    "    library_name,\n",
    "    library_ref_id,\n",
    "    log_umi_counts_filter_threshold,\n",
    ") in zip(\n",
    "    library_list,\n",
    "    library_name_list,\n",
    "    library_ref_id_list,\n",
    "    log_umi_counts_filter_threshold_list,\n",
    "):\n",
    "    print(\"Processing:\", library_name)\n",
    "    library_dir = os.path.join(\n",
    "        nanopore_project_dir, \"fastq_pass\", nano_tcr_name, library\n",
    "    )\n",
    "    umi_counts_csv = os.path.join(library_dir, \"counts\", \"umi_consensus_counts.csv\")\n",
    "    logs_dir = os.path.join(library_dir, \"logs\")\n",
    "    fastq_stats_before_filtering_txt = os.path.join(\n",
    "        logs_dir, library + \"_trimmed_stats_before_vsearch_filtering.txt\"\n",
    "    )\n",
    "    number_subreads_blast_id_csv = os.path.join(\n",
    "        logs_dir, \"merged_consensus_number_of_subreads_blast_id.csv\"\n",
    "    )\n",
    "    merged_consensus_bam_filter_log = os.path.join(\n",
    "        logs_dir, \"merged_consensus_bam_filter.log\"\n",
    "    )\n",
    "    blast_id_csv = os.path.join(logs_dir, \"merged_consensus_blast_id.csv\")\n",
    "    nt_too_short_csv = os.path.join(logs_dir, \"merged_consensus_nt_too_short.csv\")\n",
    "    region_nt_too_short_csv = os.path.join(\n",
    "        logs_dir, \"merged_consensus_region_nt_too_short.csv\"\n",
    "    )\n",
    "    nt_too_long_csv = os.path.join(logs_dir, \"merged_consensus_nt_too_long.csv\")\n",
    "    region_nt_too_long_csv = os.path.join(\n",
    "        logs_dir, \"merged_consensus_region_nt_too_long.csv\"\n",
    "    )\n",
    "    self_homology_log = os.path.join(\n",
    "        nanopore_project_dir,\n",
    "        reference_name + \"_ref_homology_out_generate_region_split_dict.log\",\n",
    "    )\n",
    "    outs_dir = os.path.join(nanopore_project_outs_dir, library + \"_\" + library_name)\n",
    "    os.mkdir(outs_dir)\n",
    "    tcr_refs_df = pd.read_csv(os.path.join(nanopore_project_dir, \"tcr_refs_df.csv\"))\n",
    "    \n",
    "    if library_ref_id:\n",
    "        library_non_n_ref_names_set = set(\n",
    "            [ref for ref in non_n_ref_names_set if library_ref_id in ref]\n",
    "        )\n",
    "        library_full_n_ref_names_set = set(\n",
    "            [ref for ref in full_n_ref_names_set if library_ref_id in ref]\n",
    "        )\n",
    "        library_cdr3j_n_ref_names_set = set(\n",
    "            [ref for ref in cdr3j_n_ref_names_set if library_ref_id in ref]\n",
    "        )\n",
    "        library_v_n_ref_names_set = set(\n",
    "            [ref for ref in v_n_ref_names_set if library_ref_id in ref]\n",
    "        )\n",
    "    else:\n",
    "        library_non_n_ref_names_set = non_n_ref_names_set\n",
    "        library_full_n_ref_names_set = full_n_ref_names_set\n",
    "        library_cdr3j_n_ref_names_set = cdr3j_n_ref_names_set\n",
    "        library_v_n_ref_names_set = v_n_ref_names_set\n",
    "\n",
    "    raw_nanopore_qual = parse_raw_nanopore_qual_from_fastq_stats(\n",
    "        fastq_stats_before_filtering_txt=fastq_stats_before_filtering_txt\n",
    "    )\n",
    "    (\n",
    "        minimal_blast_id,\n",
    "        max_too_many_nt,\n",
    "        max_too_few_nt,\n",
    "        percent_too_few_nt,\n",
    "        percent_too_many_nt,\n",
    "        percent_correct_overlap_length,\n",
    "    ) = parse_merged_consensus_bam_filter_log(\n",
    "        merged_consensus_bam_filter_log=merged_consensus_bam_filter_log\n",
    "    )\n",
    "    quantile_95_blast_id = parse_quantile_95_blast_id_from_self_homology_log(\n",
    "        self_homology_log=self_homology_log\n",
    "    )\n",
    "\n",
    "    print(\"Plotting blast_id_with_most_similar_tcr\")\n",
    "\n",
    "    plot_blast_id_with_most_similar_tcr(\n",
    "        nanopore_project_dir=nanopore_project_dir,\n",
    "        xlim_min=0.96,\n",
    "        ylim_max=blast_id_ylim_max,\n",
    "        library_ref_id=library_ref_id,\n",
    "        minimal_blast_id=minimal_blast_id,\n",
    "        quantile_95_blast_id=quantile_95_blast_id,\n",
    "        raw_nanopore_qual=raw_nanopore_qual,\n",
    "        title=library_name,\n",
    "        save_suffix=None,\n",
    "        outs_dir=outs_dir,\n",
    "    )\n",
    "\n",
    "    plot_blast_id_with_most_similar_tcr(\n",
    "        nanopore_project_dir=nanopore_project_dir,\n",
    "        xlim_min=0.99,\n",
    "        ylim_max=blast_id_ylim_max,\n",
    "        library_ref_id=library_ref_id,\n",
    "        minimal_blast_id=minimal_blast_id,\n",
    "        quantile_95_blast_id=quantile_95_blast_id,\n",
    "        raw_nanopore_qual=raw_nanopore_qual,\n",
    "        title=library_name,\n",
    "        save_suffix=\"zoom_in\",\n",
    "        outs_dir=outs_dir,\n",
    "    )\n",
    "\n",
    "    print(\"Plotting hist_subreads_per_umi_cluster\")\n",
    "    plot_hist_subreads_per_umi_cluster(\n",
    "        number_subreads_blast_id_csv=number_subreads_blast_id_csv,\n",
    "        title=library_name,\n",
    "        outs_dir=outs_dir,\n",
    "    )\n",
    "\n",
    "    print(\"Plotting box_blast_id_vs_subreads_umi_cluster\")\n",
    "    plot_box_blast_id_vs_subreads_umi_cluster(\n",
    "        number_subreads_blast_id_csv=number_subreads_blast_id_csv,\n",
    "        minimal_blast_id=minimal_blast_id,\n",
    "        title=library_name,\n",
    "        outs_dir=outs_dir,\n",
    "    )\n",
    "\n",
    "    print(\"Plotting hist_nt_too_short\")\n",
    "    plot_hist_nt_too_short(\n",
    "        nt_too_short_csv=nt_too_short_csv,\n",
    "        tcr_refs_df=tcr_refs_df,\n",
    "        max_too_few_nt=max_too_few_nt,\n",
    "        percent_too_few_nt=percent_too_few_nt,\n",
    "        title=library_name,\n",
    "        outs_dir=outs_dir,\n",
    "    )\n",
    "\n",
    "    print(\"Plotting hist_nt_too_long\")\n",
    "    plot_hist_nt_too_long(\n",
    "        nt_too_long_csv=nt_too_long_csv,\n",
    "        max_too_many_nt=max_too_many_nt,\n",
    "        percent_too_many_nt=percent_too_many_nt,\n",
    "        title=library_name,\n",
    "        outs_dir=outs_dir,\n",
    "    )\n",
    "\n",
    "    print(\"Plotting hist_percent_alignments_above_blast_id\")\n",
    "    plot_hist_percent_alignments_above_blast_id(\n",
    "        blast_id_csv=blast_id_csv,\n",
    "        minimal_blast_id=minimal_blast_id,\n",
    "        quantile_95_blast_id=quantile_95_blast_id,\n",
    "        percent_correct_overlap_length=percent_correct_overlap_length,\n",
    "        title=library_name,\n",
    "        outs_dir=outs_dir,\n",
    "    )\n",
    "\n",
    "    counts_df = pd.read_csv(umi_counts_csv)\n",
    "    counts_df = log_normalize_umi_counts(counts_df=counts_df)\n",
    "\n",
    "    print(\"Plotting umi_counts_hists\")\n",
    "    plot_umi_counts_hist(\n",
    "        counts_df=counts_df,\n",
    "        xlim_max=umi_count_xlim_max,\n",
    "        bin_size=umi_count_bin_size,\n",
    "        ylim_max=umi_count_ylim_max,\n",
    "        title=library_name,\n",
    "        plot_nb_dist_fit=True,\n",
    "        plot_percentiles=True,\n",
    "        save_suffix=None,\n",
    "        outs_dir=outs_dir,\n",
    "    )\n",
    "\n",
    "    plot_umi_counts_hist(\n",
    "        counts_df=counts_df,\n",
    "        xlim_max=umi_count_xlim_max,\n",
    "        bin_size=umi_count_bin_size,\n",
    "        ylim_max=umi_count_ylim_max,\n",
    "        title=library_name,\n",
    "        plot_nb_dist_fit=False,\n",
    "        plot_percentiles=True,\n",
    "        save_suffix=\"without_nb_dist_fit\",\n",
    "        outs_dir=outs_dir,\n",
    "    )\n",
    "\n",
    "    print(\"Plotting log_transformed_umi_counts_hists\")\n",
    "    plot_log_transformed_umi_counts_hist(\n",
    "        counts_df=counts_df,\n",
    "        xlim_max=log_umi_count_xlim_max,\n",
    "        bin_size=log_umi_count_bin_size,\n",
    "        ylim_max=log_umi_count_ylim_max,\n",
    "        custom_tcr_set=None,\n",
    "        label_custom_tcr_set=None,\n",
    "        density=True,\n",
    "        plot_normal_dist_fit=True,\n",
    "        log_umi_counts_filter_threshold=log_umi_counts_filter_threshold,\n",
    "        plot_percentiles=True,\n",
    "        most_similar_blast_id_threshold=None,\n",
    "        title=library_name,\n",
    "        save_suffix=None,\n",
    "        nanopore_project_dir=nanopore_project_dir,\n",
    "        outs_dir=outs_dir,\n",
    "    )\n",
    "\n",
    "    plot_log_transformed_umi_counts_hist(\n",
    "        counts_df=counts_df,\n",
    "        xlim_max=log_umi_count_xlim_max,\n",
    "        bin_size=log_umi_count_bin_size,\n",
    "        ylim_max=log_umi_count_ylim_max,\n",
    "        custom_tcr_set=None,\n",
    "        label_custom_tcr_set=None,\n",
    "        density=True,\n",
    "        plot_normal_dist_fit=False,\n",
    "        log_umi_counts_filter_threshold=log_umi_counts_filter_threshold,\n",
    "        plot_percentiles=True,\n",
    "        most_similar_blast_id_threshold=None,\n",
    "        title=library_name,\n",
    "        save_suffix=\"without_normal_dist_fit\",\n",
    "        nanopore_project_dir=nanopore_project_dir,\n",
    "        outs_dir=outs_dir,\n",
    "    )\n",
    "\n",
    "    plot_log_transformed_umi_counts_hist(\n",
    "        counts_df=counts_df,\n",
    "        xlim_max=log_umi_count_xlim_max,\n",
    "        bin_size=log_umi_count_bin_size,\n",
    "        ylim_max=log_umi_count_ylim_max,\n",
    "        custom_tcr_set=None,\n",
    "        label_custom_tcr_set=None,\n",
    "        density=True,\n",
    "        plot_normal_dist_fit=False,\n",
    "        log_umi_counts_filter_threshold=log_umi_counts_filter_threshold,\n",
    "        plot_percentiles=True,\n",
    "        most_similar_blast_id_threshold=most_similar_blast_id_threshold,\n",
    "        title=library_name,\n",
    "        save_suffix=\"colored_by_most_similar_blast_id\",\n",
    "        nanopore_project_dir=nanopore_project_dir,\n",
    "        outs_dir=outs_dir,\n",
    "    )\n",
    "\n",
    "    if counts_df.loc[:, \"TCR\"].isin(library_full_n_ref_names_set).any():\n",
    "        plot_log_transformed_umi_counts_hist(\n",
    "            counts_df=counts_df,\n",
    "            xlim_max=log_umi_count_xlim_max,\n",
    "            bin_size=log_umi_count_bin_size,\n",
    "            ylim_max=log_umi_count_ylim_max,\n",
    "            custom_tcr_set=library_full_n_ref_names_set,\n",
    "            label_custom_tcr_set=\"full-length synthetic\\negative control TCRs\",\n",
    "            density=True,\n",
    "            plot_normal_dist_fit=False,\n",
    "            log_umi_counts_filter_threshold=log_umi_counts_filter_threshold,\n",
    "            plot_percentiles=True,\n",
    "            most_similar_blast_id_threshold=None,\n",
    "            title=library_name,\n",
    "            save_suffix=\"colored_by_full_n_refs\",\n",
    "            nanopore_project_dir=nanopore_project_dir,\n",
    "            outs_dir=outs_dir,\n",
    "        )\n",
    "\n",
    "    if counts_df.loc[:, \"TCR\"].isin(library_cdr3j_n_ref_names_set).any():\n",
    "        plot_log_transformed_umi_counts_hist(\n",
    "            counts_df=counts_df,\n",
    "            xlim_max=log_umi_count_xlim_max,\n",
    "            bin_size=log_umi_count_bin_size,\n",
    "            ylim_max=log_umi_count_ylim_max,\n",
    "            custom_tcr_set=library_cdr3j_n_ref_names_set,\n",
    "            label_custom_tcr_set=\"CDR3-J synthetic\\negative control TCRs\",\n",
    "            density=True,\n",
    "            plot_normal_dist_fit=False,\n",
    "            log_umi_counts_filter_threshold=log_umi_counts_filter_threshold,\n",
    "            plot_percentiles=True,\n",
    "            most_similar_blast_id_threshold=None,\n",
    "            title=library_name,\n",
    "            save_suffix=\"colored_by_cdr3j_n_refs\",\n",
    "            nanopore_project_dir=nanopore_project_dir,\n",
    "            outs_dir=outs_dir,\n",
    "        )\n",
    "\n",
    "    if counts_df.loc[:, \"TCR\"].isin(library_v_n_ref_names_set).any():\n",
    "        plot_log_transformed_umi_counts_hist(\n",
    "            counts_df=counts_df,\n",
    "            xlim_max=log_umi_count_xlim_max,\n",
    "            bin_size=log_umi_count_bin_size,\n",
    "            ylim_max=log_umi_count_ylim_max,\n",
    "            custom_tcr_set=library_v_n_ref_names_set,\n",
    "            label_custom_tcr_set=\"V gene synthetic\\negative control TCRs\",\n",
    "            density=True,\n",
    "            plot_normal_dist_fit=False,\n",
    "            log_umi_counts_filter_threshold=log_umi_counts_filter_threshold,\n",
    "            plot_percentiles=True,\n",
    "            most_similar_blast_id_threshold=None,\n",
    "            title=library_name,\n",
    "            save_suffix=\"colored_by_v_n_refs\",\n",
    "            nanopore_project_dir=nanopore_project_dir,\n",
    "            outs_dir=outs_dir,\n",
    "        )\n",
    "\n",
    "    if custom_tcr_set_dict:\n",
    "        for label_custom_tcr_set, custom_tcr_set in custom_tcr_set_dict.items():\n",
    "            plot_log_transformed_umi_counts_hist(\n",
    "                counts_df=counts_df,\n",
    "                xlim_max=log_umi_count_xlim_max,\n",
    "                bin_size=log_umi_count_bin_size,\n",
    "                ylim_max=log_umi_count_ylim_max,\n",
    "                custom_tcr_set=custom_tcr_set,\n",
    "                label_custom_tcr_set=label_custom_tcr_set,\n",
    "                density=True,\n",
    "                plot_normal_dist_fit=False,\n",
    "                log_umi_counts_filter_threshold=log_umi_counts_filter_threshold,\n",
    "                plot_percentiles=True,\n",
    "                most_similar_blast_id_threshold=None,\n",
    "                title=library_name,\n",
    "                save_suffix=\"colored_by_\" + label_custom_tcr_set,\n",
    "                nanopore_project_dir=nanopore_project_dir,\n",
    "                outs_dir=outs_dir,\n",
    "            )\n",
    "\n",
    "            plot_log_transformed_umi_counts_hist(\n",
    "                counts_df=counts_df,\n",
    "                xlim_max=log_umi_count_xlim_max,\n",
    "                bin_size=log_umi_count_bin_size,\n",
    "                ylim_max=log_umi_count_ylim_max,\n",
    "                custom_tcr_set=custom_tcr_set,\n",
    "                label_custom_tcr_set=label_custom_tcr_set,\n",
    "                density=True,\n",
    "                plot_normal_dist_fit=False,\n",
    "                log_umi_counts_filter_threshold=log_umi_counts_filter_threshold,\n",
    "                plot_percentiles=False,\n",
    "                most_similar_blast_id_threshold=None,\n",
    "                title=library_name,\n",
    "                save_suffix=\"colored_by_\" + label_custom_tcr_set + \"_no_percentiles\",\n",
    "                nanopore_project_dir=nanopore_project_dir,\n",
    "                outs_dir=outs_dir,\n",
    "            )\n",
    "\n",
    "    print(\"Plotting v_gene_fold_change_over_input_barplot\")\n",
    "    plot_v_gene_fold_change_over_input_barplot(\n",
    "        counts_df=counts_df,\n",
    "        tcr_refs_df=tcr_refs_df,\n",
    "        trav_col=\"TRAV_IMGT_allele_collapsed\",\n",
    "        trbv_col=\"TRBV_IMGT_allele_collapsed\",\n",
    "        title=library_name,\n",
    "        outs_dir=outs_dir,\n",
    "    )\n",
    "\n",
    "    print(\"Plotting v_gene_fraction_missing_tcrs\")\n",
    "    plot_v_gene_fraction_missing_tcrs(\n",
    "        counts_df=counts_df,\n",
    "        tcr_refs_df=tcr_refs_df,\n",
    "        non_n_ref_names_set=library_non_n_ref_names_set,\n",
    "        region_name_col=\"TCR\",\n",
    "        trav_col=\"TRAV_IMGT_allele_collapsed\",\n",
    "        trbv_col=\"TRBV_IMGT_allele_collapsed\",\n",
    "        title=library_name,\n",
    "        outs_dir=outs_dir,\n",
    "    )\n",
    "\n",
    "    counts_df = filter_counts_on_log_umi_count_threshold(\n",
    "        counts_df=counts_df,\n",
    "        log_umi_counts_filter_threshold=log_umi_counts_filter_threshold,\n",
    "    )\n",
    "\n",
    "    plot_log_transformed_umi_counts_hist(\n",
    "        counts_df=counts_df,\n",
    "        xlim_max=log_umi_count_xlim_max,\n",
    "        bin_size=log_umi_count_bin_size,\n",
    "        ylim_max=log_umi_count_ylim_max,\n",
    "        custom_tcr_set=None,\n",
    "        label_custom_tcr_set=None,\n",
    "        density=True,\n",
    "        plot_normal_dist_fit=False,\n",
    "        log_umi_counts_filter_threshold=None,\n",
    "        plot_percentiles=True,\n",
    "        most_similar_blast_id_threshold=None,\n",
    "        title=library_name,\n",
    "        save_suffix=\"after_log_umi_count_filtering\",\n",
    "        nanopore_project_dir=nanopore_project_dir,\n",
    "        outs_dir=outs_dir,\n",
    "    )\n",
    "\n",
    "    print(\"Writing results_summary_txt\")\n",
    "    write_results_summary_txt(\n",
    "        counts_df=counts_df,\n",
    "        merged_consensus_bam_filter_log=merged_consensus_bam_filter_log,\n",
    "        region_name_col=\"TCR\",\n",
    "        library_name=library_name,\n",
    "        non_n_ref_names_set=library_non_n_ref_names_set,\n",
    "        outs_dir=outs_dir,\n",
    "    )\n",
    "\n",
    "    if plot_plates:\n",
    "        print(\"Plotting plate_umi_counts\")\n",
    "        plot_plate_umi_counts(\n",
    "            counts_df=counts_df,\n",
    "            non_n_ref_names_set=library_non_n_ref_names_set,\n",
    "            region_name_col=\"TCR\",\n",
    "            outs_dir=outs_dir,\n",
    "        )\n",
    "\n",
    "    print(\"Writing filtered_umi_count csv\")\n",
    "    counts_df.to_csv(\n",
    "        os.path.join(outs_dir, library + \"_\" + library_name + \"_filtered_umi_count.csv\")\n",
    "    )\n",
    "    print(\"\\n\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "region_nt_too_short_df = pd.read_csv(region_nt_too_short_csv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "region_nt_too_short_df.loc[:, \"region\"].value_counts()[:60]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "region_nt_too_long_df = pd.read_csv(region_nt_too_long_csv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "region_nt_too_long_df.loc[:, \"region\"].value_counts()[:60]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py310_nanopore_tcr_consensus_v3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
